source/extensions/tracers/datadog/tracer.cc

```diff
@@ -0,0 +1,126 @@
+#include "source/extensions/tracers/datadog/span.h"
+#include "source/extensions/tracers/datadog/time_util.h"
+
+#include "datadog/dict_reader.h"
+#include "datadog/error.h"
+#include "datadog/sampling_priority.h"
+#include "datadog/span_config.h"
+#include "datadog/trace_segment.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace Datadog {
+namespace {
+
+ThreadLocal::ThreadLocalObjectSharedPtr makeThreadLocalTracer(
+    datadog::tracing::TracerConfig config, Upstream::ClusterManager& cluster_manager,
+    const std::string& collector_cluster, const std::string& collector_reference_host,
+    TracerStats& tracer_stats, Event::Dispatcher& dispatcher, spdlog::logger& logger) {
+  config.logger = std::make_shared<Logger>(logger);
+  config.agent.event_scheduler = std::make_shared<EventScheduler>(dispatcher);
+  config.agent.http_client = std::make_shared<AgentHTTPClient>(
+      cluster_manager, collector_cluster, collector_reference_host, tracer_stats);
+
+  auto maybe_config = datadog::tracing::finalize_config(config);
+  if (auto* error = maybe_config.if_error()) {
+    auto prefix = "Unable to configure Datadog tracer. Tracing is now disabled. Error: ";
+    config.logger->log_error(error->with_prefix(prefix));
+    return std::make_shared<Tracer::ThreadLocalTracer>();
+  }
+
+  return std::make_shared<Tracer::ThreadLocalTracer>(*maybe_config);
+}
+
+} // namespace
+
+Tracer::ThreadLocalTracer::ThreadLocalTracer(const datadog::tracing::FinalizedTracerConfig& config)
+    : tracer(config) {}
+
+Tracer::Tracer(const std::string& collector_cluster, const std::string& collector_reference_host,
+               const datadog::tracing::TracerConfig& config,
+               Upstream::ClusterManager& cluster_manager, Stats::Scope& scope,
+               ThreadLocal::SlotAllocator& thread_local_slot_allocator)
+    : tracer_stats_(makeTracerStats(scope)),
+      thread_local_slot_(thread_local_slot_allocator.allocateSlot()) {
+  const bool allow_added_via_api = true;
+  Config::Utility::checkCluster("envoy.tracers.datadog", collector_cluster, cluster_manager,
+                                allow_added_via_api);
+
+  thread_local_slot_->set([&logger = ENVOY_LOGGER(), collector_cluster, collector_reference_host,
+                           config, &tracer_stats = tracer_stats_,
+                           &cluster_manager](Event::Dispatcher& dispatcher) {
+    return makeThreadLocalTracer(config, cluster_manager, collector_cluster,
+                                 collector_reference_host, tracer_stats, dispatcher, logger);
+  });
+}
+
+// Tracer::TracingDriver
+
+Tracing::SpanPtr Tracer::startSpan(const Tracing::Config&, Tracing::TraceContext& trace_context,
+                                   const std::string& operation_name, SystemTime start_time,
+                                   const Tracing::Decision tracing_decision) {
+  auto& thread_local_tracer = thread_local_slot_->getTyped<ThreadLocalTracer>();
```

Comment:
consider using TypedSlot for thread-safe slots.

IMO the untyped slots are tech-debt; it would be good to remove them from the public API for TheadLocalStore, but it's so hard to find the time for things like this :)

here's an example usage: source/common/rds/rds_route_config_provider_impl.h:46